pints/_optimisers/_hager_zhang_line_search.py

```diff
@@ -0,0 +1,899 @@
+        bracket = postion
+        self._minimum_alpha = bracket[0]
+        self._maximum_alpha = bracket[1]
+
+        # loop of steps L0-L3 from [1]
+        self.__line_search_looping = True
+        while self.__line_search_looping:
+
+            a = self._minimum_alpha
+            b = self._maximum_alpha
+
+            converged = self.__very_close(a, b)
+            if converged:
+                pass
+            # step L1 of line search from [1]
+            # calling secant2 as a generator
+            secant2_gen = self.__secant2(a, b)
+            wolfe_check, continue_secant2, postion = next(secant2_gen)
+            while continue_secant2 is True:
+                # print(postion)
+                yield wolfe_check, postion
+                wolfe_check, continue_secant2, postion = next(secant2_gen)
+
+            # step L2 of line search from [1]
+            # determing whether a bisection step should be preformed
+            # i.e if self.__secant_for_interval() didn't shrink the
+            # bracketing interval by the propotion self.__gamma
+            new_width = b - a
+            old_width = (self._maximum_alpha - self._minimum_alpha)
+
+            if new_width > (self.__gamma * old_width):
+                # preforming bisection
+                c = (a + b) / 2.0
+                a, b = self.__update(a=a, b=b, c=c)
+
+            # step L3 of the line search from [1]
+            # updating bracket interval
+            self._minimum_alpha = a
+            self._maximum_alpha = b
+
+    def tell(self, reply):
+        """ See :meth:`Optimiser.tell()`. """
+
+        # print('in tell')
+
+        # Check ask-tell pattern
+        if not self._ready_for_tell:
+            raise Exception('ask() not called before tell()')
+        self._ready_for_tell = False
+
+        # Unpack reply
+        self.__reply_f_and_dfdx = reply
+        proposed_f, proposed_dfdx = reply[0]
+        proposed_f = proposed_f
+        proposed_dfdx = np.asarray(proposed_dfdx)
+
+        # We need the evaluation of the gradients before we can start the BFGS,
+        # the first tell gets these.
+        if self._current_f is None:
+
+            # Move to proposed point
+            self._current = self._proposed
+            self._current_f = np.asarray(proposed_f)
+            self._current_dfdx = np.asarray(proposed_dfdx)
+
+        elif self.wolfe_check is True:
+            # Checking if exact wolfe conditions are meet.
+            proposed_grad = np.matmul(np.transpose(proposed_dfdx), self.__px)
+
+            wolfe_curvature = (proposed_grad >= self._c2 *
+                               np.matmul(np.transpose(self._current_dfdx),
+                                         self.__px))
+
+            exact_wolfe_suff_dec = (self._c1 *
+                                    np.matmul(np.transpose(self._current_dfdx),
+                                              self.__px)
+                                    >= proposed_f - self._current_f)
```

Comment:
`np.matmul(np.transpose(self._current_dfdx), self.__px)` seems to be being repeated three times here? I'd precalculate it and then reuse.